### PR TITLE
Remove TODO from 2015 to detect host refresh rate/max fps for Mir-on-X client.

### PR DESCRIPTION
**Pull request by pillowtrucker**
There's a few sticky things here:
- First of all, I'm adding a new dependency. However, xcb-randr should already exist anywhere xcb is installed, as far as I know. On my distribution this is not even a separate package. I've added the relevant stanza for debian/ubuntu in the control file (commented out for now), but, if I remember debian packaging correctly, this won't apply until the version is bumped in the changelog. 
- I'm following the general trend of ignoring xcb errors. They're usually non-fatal.
- If the randr extension version of the host X11 is <1.2, this will probably either not work or return nonsense. OTOH, if someone is doing development or testing of Mir-on-X on an Xorg that doesn't support 1.2+, they're probably doing it deliberately.
- More importantly, should I also follow [this guide](https://github.com/MirServer/mir/blob/c1d267021c9c13798d161d7ca73aad9bd4b93bbe/doc/sphinx/dso_versioning_guide.md?plain=1#L106) to make a versioned `mgx::DisplayConfiguration::build_output` ? Would I also have to add it to the relevant symbols.map, first ?

### CODE REVIEW

**Comment by pillowtrucker** [comment]
The new build errors are actually interesting - it seems that many targets aren't capable of building c++20, let alone c++23

**Comment by pillowtrucker** [comment]
xcb would not, in fact, be the most arcane part of this PR.

**Comment by AlanGriffiths** [comment]
> it seems that many targets aren't capable of building c++20, let alone c++23, which I assumed is the standard we should aim to conform to, based on the cmake files..
> Should I just replace std::format with something more verbose ?

The "oldest" target we support is the one corresponding to the base in our various snaps (which is currently `core22` - corresponding to 22.04LTS). The compiler support for c++23 is mostly there but, as you encountered, the library support is lacking in places.

We aim to transition to `core24`/24.04LTS once these are released and supported.

**Comment by AlanGriffiths** [comment]
> * If the randr extension version of the host X11 is <1.2, this will probably either not work or return nonsense. OTOH, if someone is doing development or testing of Mir-on-X on an Xorg that doesn't support 1.2+, they're probably doing it deliberately.

We don't make any effort to support development on anything earlier than our current snap base (currently `core22`) - which is Ubuntu 22.04LTS. We do however, expect snaps (such as Ubuntu Frame) to run on supported versions of Ubuntu (so, that currently includes 14.04LTS which is still on ESM).

But even 14.04 (trusty) should be fine:

```
$ rmadison libxcb-randr0-dev
 libxcb-randr0-dev | 1.10-2ubuntu1      | trusty         | amd64, arm64, armhf, i386, powerpc, ppc64el
 libxcb-randr0-dev | 1.11.1-1ubuntu1    | xenial         | amd64, arm64, armhf, i386, powerpc, ppc64el, s390x
 libxcb-randr0-dev | 1.13-1             | bionic         | amd64, arm64, armhf, i386, ppc64el, s390x
 libxcb-randr0-dev | 1.13-2~ubuntu18.04 | bionic-updates | amd64, arm64, armhf, i386, ppc64el, s390x
 libxcb-randr0-dev | 1.14-2             | focal          | amd64, arm64, armhf, i386, ppc64el, riscv64, s390x
 libxcb-randr0-dev | 1.14-3ubuntu3      | jammy          | amd64, arm64, armhf, i386, ppc64el, riscv64, s390x
 libxcb-randr0-dev | 1.15-1             | lunar          | amd64, arm64, armhf, i386, ppc64el, riscv64, s390x
 libxcb-randr0-dev | 1.15-1             | mantic         | amd64, arm64, armhf, i386, ppc64el, riscv64, s390x
 libxcb-randr0-dev | 1.15-1             | noble          | amd64, arm64, armhf, i386, ppc64el, riscv64, s390x
```

File: debian/control
Comment (AlanGriffiths) [inline debian/control:48]: There's no need for this. And the comment would date rapidly

File: src/platforms/x11/x11_resources.h
Comment (AlanGriffiths) [inline src/platforms/x11/x11_resources.h:35]: I don't think any of these changes are needed in _this header_. Those additional headers needed by the implementation should be added to the implementation file.

File: src/platforms/x11/x11_resources.cpp
Comment (AlanGriffiths) [inline src/platforms/x11/x11_resources.cpp:123]:
```suggestion
        mir::log_debug("%s", log_msg.c_str());
```

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
Looks mostly good with a few tweaks.

The snaps action failing should be addressed by #3137, so I recommend rebasing on main once that lands.

It does seem inelegant to use two sources (`if __has_include(<format>)` and `#ifdef __cpp_lib_format`) for the one truth, but I' wouldn't block on it.

File: src/platforms/x11/x11_resources.cpp
Comment (pillowtrucker) [reply src/platforms/x11/x11_resources.cpp]: The existence of log_debug makes the whole thing  much much simpler.

**Comment by pillowtrucker** [comment]
My apologies. The `#ifdef` inconsistency and header misplacement was purely due to blindly following clangd's preferences - it tries to be clever and take the tested expression into consideration while syntax-checking, but doesn't really store enough context to keep it consistent. 
The existence of the mir loggers with the standard format specifier significantly simplifies the logging code, so I've cleaned all of that up and rebased onto main.
It looks like some of the CI jobs are a little bit confused now, though..

File: src/platforms/x11/x11_resources.cpp
Comment (AlanGriffiths) [inline src/platforms/x11/x11_resources.cpp:29]: AFAICS you've not added a requirement for any of these?

File: src/platforms/x11/x11_resources.cpp
Comment (AlanGriffiths) [inline src/platforms/x11/x11_resources.cpp:17]: Spurious new whitespace

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
Much cleaner, thanks. A couple of nits

**Comment by pillowtrucker** [comment]
All fixed. I think the CI failures are still transient and unrelated to the code. (hopefully...)

**Review by AlanGriffiths — APPROVED** [review]

**Pull request merged**
